File: cpp/src/remaining_methods_3310.cpp
// https://leetcode.com/problems/remove-methods-from-project

#include "test.hpp"

#include <vector>

    std::vector<int> remainingMethods(int n_, int k_, const std::vector<std::vector<int>>& invocations) {
        using namespace std;
        size_t n = static_cast<size_t>(n_);
        size_t k = static_cast<size_t>(k_);
        vector<vector<size_t>> graph(n);
        vector<int> in_degree(n, 0);
        for (const auto& edge : invocations) {
            size_t u = static_cast<size_t>(edge[0]);
            size_t v = static_cast<size_t>(edge[1]);
            graph[u].push_back(v);
            in_degree[v]++;
        }
        vector<bool> visited(n, false);
        vector<size_t> stack;
        stack.push_back(k);
        visited[k] = true;
        while (!stack.empty()) {
            size_t node = stack.back();
            stack.pop_back();
            for (size_t neighbor : graph[node]) {
                in_degree[neighbor]--;
                if (!visited[neighbor]) {
                    visited[neighbor] = true;
                    stack.push_back(neighbor);
                }
            }
        }
        vector<int> result;
        bool external_reference_found = false;
        for (size_t i = 0; i < n; ++i) {
            if (visited[i]) {
                if (in_degree[i] > 0) {
                    external_reference_found = true;
                    break;
                }
            } else {
                result.push_back(static_cast<int>(i));
            }
        }
        if (!external_reference_found) {
            return result;
        }
        result.clear();
        for (size_t i = 0; i < n; ++i) {
            result.push_back(static_cast<int>(i));
        }
        return result;
    }

TEST(official1) {
    check_eq(
        std::vector<int>{0, 1, 2, 3},
        remainingMethods(4, 1, {{1, 2}, {0, 1}, {3, 2}})
    );
}

TEST(official2) {
    check_eq(
        std::vector<int>{3, 4},
        remainingMethods(5, 0, {{1, 2}, {0, 2}, {0, 1}, {3, 4}})
    );
}

TEST(official3) {
    check_eq(
        std::vector<int>{},
        remainingMethods(3, 2, {{1, 2}, {0, 1}, {2, 0}})
    );
}

File: rust/src/remaining_methods_3310.rs
// https://leetcode.com/problems/remove-methods-from-project

pub fn remaining_methods(n: i32, k: i32, invocations: Vec<Vec<i32>>) -> Vec<i32> {
    let components = connected_components(n as usize, &invocations);
    let k_component = components[k as usize];
    let k_component_size = components.iter().filter(|&&c| c == k_component).count();
    
    let mut graph = vec![vec![]; n as usize];
    for edge in invocations {
        let u = edge[0] as usize;
        let v = edge[1] as usize;
        graph[u].push(v);
    }
    
    //dfs
    let mut stack = vec![k as usize];
    let mut counter = 0;
    let mut visited = vec![false; n as usize];
    visited[k as usize] = true;
    while let Some(node) = stack.pop() {
        counter += 1;
        for &neighbor in &graph[node] {
            debug_assert!(components[neighbor] == k_component);
            if !visited[neighbor] {
                visited[neighbor] = true;
                stack.push(neighbor);
            }
        }
    }

    if k_component_size == counter {
        (0..n).filter(|&x| components[x as usize] != k_component).collect()
    } else {
        (0..n).collect()
    }

}

// https://dsar.rantai.dev/docs/part-iii/chapter-14/
fn connected_components(vertice_count: usize, roads: &Vec<Vec<i32>>) -> Vec<usize> {
    let mut parents: Vec<usize> = (0..vertice_count).collect();
    let slice = parents.as_mut_slice();
    for edge in roads {
        add_component(edge, slice);
    }
    for i in 0..vertice_count {
        find_component(i, &mut parents);
    }
    parents
}

fn add_component(new_component: &[i32], parents: &mut [usize]) -> usize {
    debug_assert!(new_component.len() >= 2);
    let component = find_component( into_usize(*new_component.first().unwrap()), parents);
    for &i in new_component[1..].iter() {
        parents[find_component(into_usize(i), parents)] = component;
    }
    component
}

fn find_component(x: usize, parent: &mut [usize]) -> usize {
    let mut root = x;
    while parent[root] != root {
        root = parent[root];
    }
    // Path compression
    let mut cur = x;
    while parent[cur] != root {
        let next = parent[cur];
        parent[cur] = root; 
        cur = next;
    }
    root
}

fn into_usize(input: i32) -> usize {
    usize::try_from(input).expect("Input can't be negative")
}


#[cfg(test)]
mod tests {
    use super::remaining_methods;

    #[test]
    fn official1() {
        let n = 4;
        let k = 1;
        let invocations = vec![
            vec![1, 2],
            vec![0, 1],
            vec![3, 2],
        ];
        assert_eq!(vec![0, 1, 2, 3], remaining_methods(n, k, invocations));
    }

    #[test]
    fn official2() {
        let n = 5;
        let k = 0;
        let invocations = vec![
            vec![1, 2],
            vec![0, 2],
            vec![0, 1],
            vec![3, 4],
        ];
        assert_eq!(vec![3, 4], remaining_methods(n, k, invocations));
    }

    #[test]
    fn official3() {
        let n = 3;
        let k = 2;
        let invocations = vec![
            vec![1, 2],
            vec![0, 1],
            vec![2, 0],
        ];
        let expected: Vec<i32> = vec![];
        assert_eq!(expected, remaining_methods(n, k, invocations));
    }
}
